src/apps/bm_devkit/sdi12_example/user_code/user_code.cpp

```diff
@@ -0,0 +1,196 @@
+  // Inquire Sensor address
+  sondeEXO3sSensor.inquire_cmd();
+  vTaskDelay(pdMS_TO_TICKS(100));
+  // Inquire Identification
+  sondeEXO3sSensor.identify_cmd();
+}
+
+void loop(void) {
+  // get sensor readings
+  sondeEXO3sSensor.measure_cmd();
+  // accumulate list of these samples
+  EXO3sample sens = sondeEXO3sSensor.getLatestSample();
+  samples[current_sample_index] = sens;
+
+  // Get the RTC if available
+  RTCTimeAndDate_t time_and_date = {};
+  rtcGet(&time_and_date);
+  char rtcTimeBuffer[32];
+  rtcPrint(rtcTimeBuffer, &time_and_date);
+
+  // log values in the sd card
+  bm_fprintf(0, "exo3s_sonde_raw.log", USE_TIMESTAMP, "tick: %llu, rtc: %s, data --- "
+                                       "temp_sensor: %.3f C, "
+                                       "sp_cond: %.3f uS/cm, "
+                                       "pH: %.3f, "
+                                       "pH: %.3f mV, "
+                                       "DO: %.3f Percent Sat, "
+                                       "DO: %.3f mg/L, "
+                                       "turbidity: %.3f NTU, "
+                                       "wiper pos: %.3f V, "
+                                       "depth: %.3f m, "
+                                       "power supply: %.3f V"
+                                       "\n", uptimeGetMs(), rtcTimeBuffer, sens.temp_sensor, sens.sp_cond, sens.pH, sens.pH_mV,
+             sens.dis_oxy, sens.dis_oxy_mg, sens.turbidity, sens.wiper_pos, sens.depth, sens.power) ;
+  // print to spotter console
+  bm_printf(0, "[payload] | tick: %llu, rtc: %s, data --- "
+           "temp_sensor: %.3f C, "
+           "sp_cond: %.3f uS/cm, "
+           "pH: %.3f, "
+           "pH: %.3f mV, "
+           "DO: %.3f Percent Sat, "
+           "DO: %.3f mg/L, "
+           "turbidity: %.3f NTU, "
+           "wiper pos: %.3f V, "
+           "depth: %.3f m, "
+           "power supply: %.3f V"
+           "\n", uptimeGetMs(), rtcTimeBuffer, sens.temp_sensor, sens.sp_cond, sens.pH, sens.pH_mV,
+        sens.dis_oxy, sens.dis_oxy_mg, sens.turbidity, sens.wiper_pos, sens.depth, sens.power) ;
+  // print to console,
+  // max console output size is 128, therefore will split the below into 2 printf statements
+  printf("[exo3-sonde-sdi12] | tick: %llu, rtc: %s, data --- "
+         "temp_sensor: %.3f C, "
+         "sp_cond: %.3f uS/cm, "
+         "pH: %.3f, \n", uptimeGetMs(), rtcTimeBuffer, sens.temp_sensor, sens.sp_cond, sens.pH);
+  printf("pH: %.3f mV, "
+         "DO: %.3f Percent Sat, "
+         "DO: %.3f mg/L, "
+         "turbidity: %.3f NTU, "
+         "wiper pos: %.3f V, "
+         "depth: %.3f m, "
+         "power supply: %.3f V\n", sens.pH_mV, sens.dis_oxy, sens.dis_oxy_mg, sens.turbidity, sens.wiper_pos, sens.depth, sens.power);
+
+  current_sample_index++;
+  if (current_sample_index == sample_count){
+    transmit_samples();
+    current_sample_index = 0;
+  }
+  vTaskDelay(pdMS_TO_TICKS(2000));
```

Comment:
The LED code below is non-blocking, whereas this delay blocks `loop`. That means that LED1 won't actually pulse green every second as the code implies. Also LED2 won't pulse green on payload input. Both will probably pulse every 2 seconds.

It's fine as is. However, if you have time, it would be an improvement to make this sonde interaction also non-blocking. The steps would be:
- Delete the call to `vTaskDelay`.
- Extract lines 98 to 156 (the whole sonde interaction) into its own function named something like `sample_exo3_sonde`
- Add a `#define EXO3_READING_PERIOD_MS 2000` (or similar name) up around line 20.
- At the top of `loop` do something like this:

```C++
static uint64_t last_exo3_reading_uptime = uptimeGetMs();
if (uptimeGetMs() - last_exo3_reading_uptime >= EXO3_READING_PERIOD_MS) {
  last_exo3_reading_uptime = uptimeGetMs();
  sample_exo3_sonde();
}
```

Reply:
@towynlin I shall delete the 2 second delay. I had added that initially for debug purposes and is not really needed for the sensor to do its thing! Thanks for catching that

Reply:
Got it. If we don't delay, does the sensor only provide readings on a certain cadence or would that gather lots of readings really quickly, leading to our sending too much data via `spotter_tx_data`?

Reply:
The sensor does not rely on the delay. It takes a long time to make a measurement and send it over (62 SECONDS). So we have about 40 bytes of data each time we inquire a measurement. I have grouped 3 such measurements (samples) together and tx it to the spotter along with some RTC timestamp. 
Evan mentioned we have a 311 bytes limit for `spotter_tx_data`. So 29 header bytes for the `BMBinData1` message type + 40 *3 bytes sensor data, 149 bytes  < 311 bytes!

Reply:
Forgot to add - the sensor does a measurement and sends it over only after we request it to (sending a measure command), and does not measure new data automatically.